website/registrations/utils.py

```diff
@@ -0,0 +1,349 @@
+            'column_index': kwargs['column_index'],
+            'row_index': kwargs['row_index'],
+            'missing': kwargs.get('missing', False),
+            'invalid': kwargs.get('invalid', False)
+        })
+
+    def validate_csv_header_list(self):
+        expected_headers = self.validations.keys()
+        actual_headers = self.headers
+        diff = set(expected_headers) - set(actual_headers)
+        if len(diff):
+            raise ValidationError('Invalid csv headers: {}'.format(','.join(diff)))
+
+    def get_parsed(self):
+        parsed = []
+        for row in self.rows:
+            parsed.append({'csv_raw': row.get_raw_value(), 'csv_parsed': row.get_parsed_value()})
+        return {'schema_id': self.schema_id, 'registrations': parsed}
+
+    def validate(self):
+        self.validate_csv_header_list()
+        for row in self.rows:
+            row.validate()
+
+class Row():
+    def __init__(self, row_dict, validations, log_error):
+        self.row_dict = row_dict
+        self.cells = [Cell(header.lower(),
+                           value, validations[header.lower()],
+                           functools.partial(log_error, column_index=column_index))
+                           for column_index, (header, value) in enumerate(row_dict.items())]
+
+    def to_json(self):
+        json_value = {}
+        for cell in self.cells:
+            json_value.update(cell.get_parsed_value())
+        return json.dumps(json_value)
+
+    def get_metadata(self):
+        parsed_metadata = {}
+        for cell in self.cells:
+            if cell.is_metadata:
+                parsed_metadata.update(cell.get_parsed_value())
+        return parsed_metadata
+
+    def get_registration_responses(self):
+        parsed_responses = {}
+        for cell in self.cells:
+            if not cell.is_metadata:
+                parsed_responses.update(cell.get_parsed_value())
+        return parsed_responses
+
+    def get_parsed_value(self):
+        return {'metadata': self.get_metadata(),
+                'registration_responses': self.get_registration_responses()}
+
+    def get_raw_value(self):
+        raw_value = []
+        for cell in self.cells:
+            raw_value.append(cell.get_raw_value())
+        return ','.join(raw_value)
+
+    def validate(self):
+        for cell in self.cells:
+            cell.validate()
+
+class Cell():
+    @property
+    def is_metadata(self):
+        return self.header in METADATA_FIELDS.keys()
+
+    def __init__(self, header, value, validations, log_error):
+        self.header = header
+        self.value = value
+        self.validations = validations
+        self.field = Cell.field_instance_for(self.header.lower(),
+                                             value=self.value,
+                                             validations=self.validations,
+                                             log_error=functools.partial(log_error, header=self.header))
```

Comment:
If `value`, `validations` and `log_error` are all required (they are already required by the constructor), then maybe they shouldn't be key-worded arguments but positional arguments.

Reply:
With `log_error` being a partial function, it's more readable to go with `kwargs` for readability purposes.
When logging cell errors, `self.log_error(invalid=True)` is less meaningful than `self.log_error(True)`

Reply:
My comment applies to `field_instance_for`, not `functools.partial`. if `value`, `validations` and `log_error` are kwargs, that means they are optional. This goes hand in hand with my next comment below for the invocation of constructors of different `Field`s.